src/RevCPU.cc

```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
-                        Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
-        Thread->SetState(ThreadState::BLOCKED);
-        // -- 3a.
-        BlockedThreads.emplace(Thread->GetThreadID());
-
-        // -- 3b.
-        AssignedThreads.at(ProcID).erase(Thread->GetThreadID());
-        
-        if( AssignedThreads.at(ProcID).empty() ){
-          Enabled[ProcID] = false;
-        }
+      output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 "on Core %" PRIu32 " is BLOCKED\n", ThreadID, ProcID);
 
-      }
+      // Set its state to BLOCKED
+      Thread->SetState(ThreadState::BLOCKED);
+
+      // Add it to BlockedThreads
+      BlockedThreads.emplace_back(std::move(Thread));
       break;
-    case ThreadState::START: // Should never happen
-      output.verbose(CALL_INFO, 99, 1, "A new thread with ID = %" PRIu32 " was found on Core %" PRIu32, Thread->GetThreadID(), ProcID);
+    case ThreadState::START:
+      // A new thread was created
+      output.verbose(CALL_INFO, 99, 1, "A new thread with ID = %" PRIu32 " was found on Core %" PRIu32, Thread->GetID(), ProcID);
 
       // Mark it ready for execution
       Thread->SetState(ThreadState::READY);
 
-      // Add it to the Thread map
-      Threads.emplace(Thread->GetThreadID(), Thread);
+      // Add it to the ReadyThreads so it is scheduled
+      ReadyThreads.emplace_back(std::move(Thread));
 
-      // Add it to the thread queue to be scheduled
-      ThreadQueue.emplace_back(Thread->GetThreadID());
       break;
 
     case ThreadState::RUNNING:
-      output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", Thread->GetThreadID(), ProcID);
+      output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", ThreadID, ProcID);
       break;
 
     case ThreadState::READY:
       // If this happens we are not setting state when assigning thread somewhere
       output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is assigned but is in READY state... This is a bug\n",
-                    Thread->GetThreadID(), ProcID);
+                    ThreadID, ProcID);
       break;
     default: // Should DEFINITELY never happen
-      output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug\n",
-                    Thread->GetThreadID(), ProcID);
+      output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug.\n%s\n",
+                    ThreadID, ProcID, Thread->to_string().c_str());
       break;
     }
-    // Pop the thread that changed state
-    Procs[ProcID]->GetThreadsThatChangedState().pop();
+    // If the Proc does not have any threads assigned to it, there is no work to do, it is disabled
+    if( Procs[ProcID]->HasNoBusyHarts() ){
+      Enabled[ProcID] = false;
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
```

Comment:
At some point we may need to create Current vs. Next state for threads, but I think given the separate ClockTick functions in RevProc and RevCPU, we should be OK as long as we are careful in how the updates are done

Reply:
That seems like something that is definitely possible to achieve with the current implementation (ie. we haven't coded ourselves into a corner) so if it's cool with you I'll leave it as is in this PR and down the road we can revisit